### PR TITLE
feature/DW 411 access by permisisons by term

### PR DESCRIPTION
**Pull request by martinyde**
https://jira.itkdev.dk/browse/DW-411

- DW-411: Added gitignore
- DW-411: Added os2forms_webform_list module
- DW-411: Added os2forms_permissions_by_term module

### CODE REVIEW

File: web/modules/custom/os2forms_webform_list/src/customWebformEntityListBuilder.php [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    /** @var \Drupal\webform\WebformInterface $webforms */
+    $webforms = $this->getStorage()->loadMultiple();
+    foreach ($webforms as $webform) {
+      $access = $webform->access('update');
+      if($access) {
+        $webform_ids_permissions_by_term[] = $webform->id();
+      }
+    }
+
+    $query->condition('id', $webform_ids_permissions_by_term, 'IN');
```
Comment (rimi-itk) [inline web/modules/custom/os2forms_webform_list/src/customWebformEntityListBuilder.php:94]: You should have only these lines plus a call to `parent::getQuery()` in your implementation of `getQuery`, i.e.

```php
  protected function getQuery($keys = '', $category = '', $state = '') {
    $query = parent::getQuery($keys, $category, $state);

    // Setup a required condition for the list builder to respect webform update access.
    …

    return $query;
  }
```

File: web/modules/custom/os2forms_permissions_by_term/src/Helper/Helper.php [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+      case 'submission_page':
+      case 'submission_view_any':
+      case 'submission_view_own':
+      case 'submission_purge_any':
+        // Allow access if no term is set for the form or a webform term match the users term.
+        if (empty($webformPermissionsByTerm) || !empty(array_intersect($webformPermissionsByTerm, $userTerms))) {
+          return AccessResult::neutral();
+        } else {
+          return AccessResult::forbidden();
+        }
```
Comment (rimi-itk) [inline web/modules/custom/os2forms_permissions_by_term/src/Helper/Helper.php:159]: Should be written as 

```php
return empty($webformPermissionsByTerm) || !empty(array_intersect($webformPermissionsByTerm, $userTerms)
  ? AccessResult::neutral()
  : AccessResult::forbidden();
```

File: web/modules/custom/os2forms_permissions_by_term/src/Helper/Helper.php [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+  public function webformAccess(WebformInterface $webform, $operation, AccountInterface $account) {
+    $user = $this->entityTypeManager->getStorage('user')->load($account->id());
+    $userTerms = $this->accessStorage->getPermittedTids($user->id(), $user->getRoles());
+    $webformPermissionsByTerm = $webform->getThirdPartySetting('os2forms_permissions_by_term', 'settings');
+    switch ($operation) {
+      case 'view':
+        // We don't use permission by term to determine access to the actual webform.
+        // This could probably be removed, but is left in to show we are aware of this operation.
+        return AccessResult::neutral();
+        break;
```
Comment (rimi-itk) [inline web/modules/custom/os2forms_permissions_by_term/src/Helper/Helper.php:144]: Don't use `break` after `return`.

**Review by rimi-itk — CHANGES_REQUESTED** [review]

File: web/modules/custom/os2forms_webform_list/os2forms_webform_list.module [hunk 4]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+/**
+ * Implements hook_entity_type_alter().
+ */
+function os2forms_webform_list_entity_type_alter(array &$entity_types) {
+  /** @var $entity_types \Drupal\Core\Entity\EntityTypeInterface[] */
+  // Define a new list builder class.
+  $entity_types['webform']->setListBuilderClass('Drupal\os2forms_webform_list\customWebformEntityListBuilder');
```
[the remark this answers is not in the document]
Comment (martinyde) [reply web/modules/custom/os2forms_webform_list/os2forms_webform_list.module:9]: The proposed change results in :
`Drupal\Component\Plugin\Exception\InvalidPluginDefinitionException: The "webform" entity type did not specify a  
                                                list_builder handler. in Drupal\Core\Entity\EntityTypeManager->getHandler() `

[the remark this answers is not in the document]
Comment (rimi-itk) [reply web/modules/custom/os2forms_webform_list/os2forms_webform_list.module:9]: Apparently, I assume that `customWebformEntityListBuilder` has been renamed to `CustomWebformEntityListBuilder` …

File: web/modules/custom/os2forms_webform_list/src/customWebformEntityListBuilder.php [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+use Drupal\webform\WebformEntityListBuilder;
+use Drupal\webform\WebformInterface;
+
+/**
+ * Defines a class to build a listing of webform entities.
+ *
+ * @see \Drupal\webform\Entity\Webform
+ */
+class customWebformEntityListBuilder extends WebformEntityListBuilder {
```
[the remark this answers is not in the document]
Comment (martinyde) [reply web/modules/custom/os2forms_webform_list/src/customWebformEntityListBuilder.php]: Fixed

File: web/modules/custom/os2forms_webform_list/src/customWebformEntityListBuilder.php (hunk 1 above)
Comment (martinyde) [reply web/modules/custom/os2forms_webform_list/src/customWebformEntityListBuilder.php]: Fixed

File: web/modules/custom/os2forms_permissions_by_term/src/Helper/Helper.php (hunk 3 above)
Comment (martinyde) [reply web/modules/custom/os2forms_permissions_by_term/src/Helper/Helper.php]: Fixed

File: web/modules/custom/os2forms_permissions_by_term/src/Helper/Helper.php (hunk 2 above)
Comment (martinyde) [reply web/modules/custom/os2forms_permissions_by_term/src/Helper/Helper.php]: Fixed

File: web/modules/custom/os2forms_webform_list/os2forms_webform_list.module (hunk 4 above)
[the remark this answers is not in the document]
Comment (martinyde) [reply web/modules/custom/os2forms_webform_list/os2forms_webform_list.module]: Wont fix

File: web/modules/custom/os2forms_permissions_by_term/src/Helper/Helper.php [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+      '#title' => 'Webform access',
+      '#tree' => TRUE,
+      '#weight' => -99,
+    ];
+
+    $form['os2forms_permissions_by_term']['os2forms_access'] = [
+      '#type' => 'checkboxes',
+      '#required' => TRUE,
+      '#title' => t('Access'),
+      '#default_value' => 'settings' === $hook ? $defaultSettings : [],
```
Comment (rimi-itk) [inline web/modules/custom/os2forms_permissions_by_term/src/Helper/Helper.php:104]: Using `'#default_value' => $defaultSettings ?? []` here will make more sense (to me) as `$defaultSettings` is set only if `'settings' === $hook` (cf. line 86–90.

Alternatively, add `$defaultSettings = []` in line 85 and use `'#default_value' => $defaultSettings` here.

**Review by rimi-itk — CHANGES_REQUESTED** [review]

File: web/modules/custom/os2forms_permissions_by_term/src/Helper/Helper.php (hunk 6 above)
Comment (martinyde) [reply web/modules/custom/os2forms_permissions_by_term/src/Helper/Helper.php]: Fixed

**Review by rimi-itk — APPROVED** [review]

**Pull request merged**